### PR TITLE
updates to verify

### PR DESCRIPTION
**Pull request by nitro-neal**
Addressing:
https://github.com/TBD54566975/web5-js/issues/420

Made sure we are explicitly checking these:

   * exp MUST represent the expirationDate property, encoded as a UNIX timestamp (NumericDate).
   * iss MUST represent the issuer property of a verifiable credential or the holder property of a verifiable presentation.
   * nbf MUST represent issuanceDate, encoded as a UNIX timestamp (NumericDate).
   * jti MUST represent the id property of the verifiable credential or verifiable presentation.
   * sub MUST represent the id property contained in the credentialSubject.

updated the iss to be the did's id and not the vcDataModel issuer

Examples:
vcJwt:

```
eyJ0eXAiOiJKV1QiLCJhbGciOiJFZERTQSIsImtpZCI6ImRpZDpkaHQ6aWJxdXhkNWhqZjhhOXlid3o2emN1d2U1ejhtbnpoZmF1Nmp3cms1OWlwZmlycnVvano5eSMwIn0.eyJ2YyI6eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSJdLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiVEJEZXZlbG9wZXJDcmVkZW50aWFsIl0sImlkIjoidXJuOnV1aWQ6MjZmZTQ4MGQtMjFjMS00Nzc0LWI1YjctNjQ0NDc3NzcxMjNiIiwiaXNzdWVyIjoiZGlkOmRodDppYnF1eGQ1aGpmOGE5eWJ3ejZ6Y3V3ZTV6OG1uemhmYXU2andyazU5aXBmaXJydW9qejl5IiwiaXNzdWFuY2VEYXRlIjoiMjAyNC0wMi0yMVQyMzoxMTo0M1oiLCJjcmVkZW50aWFsU3ViamVjdCI6eyJpZCI6ImRpZDpkaHQ6aWJxdXhkNWhqZjhhOXlid3o2emN1d2U1ejhtbnpoZmF1Nmp3cms1OWlwZmlycnVvano5eSIsInVzZXJuYW1lIjoibml0cm8ifX0sIm5iZiI6MTcwODU1NzEwMywianRpIjoidXJuOnV1aWQ6MjZmZTQ4MGQtMjFjMS00Nzc0LWI1YjctNjQ0NDc3NzcxMjNiIiwiaXNzIjoiZGlkOmRodDppYnF1eGQ1aGpmOGE5eWJ3ejZ6Y3V3ZTV6OG1uemhmYXU2andyazU5aXBmaXJydW9qejl5Iiwic3ViIjoiZGlkOmRodDppYnF1eGQ1aGpmOGE5eWJ3ejZ6Y3V3ZTV6OG1uemhmYXU2andyazU5aXBmaXJydW9qejl5IiwiaWF0IjoxNzA4NTU3MTAzfQ.FeL8YtNltBjbNSN_nZJCkf1VnA_S_nkHY9jIihAL27kmuYHUeKvCYjXTsIuWBJnTc0AG9v35vRQAnQqTc4lXAA
```

Decoded JWT:
```
{
  "vc": {
    "@context": [
      "https://www.w3.org/2018/credentials/v1"
    ],
    "type": [
      "VerifiableCredential",
      "TBDeveloperCredential"
    ],
    "id": "urn:uuid:26fe480d-21c1-4774-b5b7-64447777123b",
    "issuer": "did:dht:ibquxd5hjf8a9ybwz6zcuwe5z8mnzhfau6jwrk59ipfirruojz9y",
    "issuanceDate": "2024-02-21T23:11:43Z",
    "credentialSubject": {
      "id": "did:dht:ibquxd5hjf8a9ybwz6zcuwe5z8mnzhfau6jwrk59ipfirruojz9y",
      "username": "nitro"
    }
  },
  "nbf": 1708557103,
  "jti": "urn:uuid:26fe480d-21c1-4774-b5b7-64447777123b",
  "iss": "did:dht:ibquxd5hjf8a9ybwz6zcuwe5z8mnzhfau6jwrk59ipfirruojz9y",
  "sub": "did:dht:ibquxd5hjf8a9ybwz6zcuwe5z8mnzhfau6jwrk59ipfirruojz9y",
  "iat": 1708557103
}
```

ParsedJwt (VcDataModel):

```
VerifiableCredential {
  vcDataModel: {
    '@context': [ 'https://www.w3.org/2018/credentials/v1' ],
    type: [ 'VerifiableCredential', 'TBDeveloperCredential' ],
    id: 'urn:uuid:26fe480d-21c1-4774-b5b7-64447777123b',
    issuer: 'did:dht:ibquxd5hjf8a9ybwz6zcuwe5z8mnzhfau6jwrk59ipfirruojz9y',
    issuanceDate: '2024-02-21T23:11:43Z',
    credentialSubject: {
      id: 'did:dht:ibquxd5hjf8a9ybwz6zcuwe5z8mnzhfau6jwrk59ipfirruojz9y',
      username: 'nitro'
    }
  }
}
```

image

## Example processing rules:

jti
jti: 123
payload.vc.id: 123
valid? :white_check_mark:

jti: 123
payload.vc.id: undefined
valid? :white_check_mark:

jti: 123
payload.vc.id: 321
valid? :x:

jti: 123
payload.vc.id: “” // empty string
valid? :x:

### CODE REVIEW

**Review by decentralgabe — APPROVED** [review]
lgtm

**Review by mistermoe — COMMENTED** [review]
@nitro-neal i have vectors [here](https://github.com/TBD54566975/web5-go/blob/main/vc/vcjwt_test.go)that may be worth using

**Comment by nitro-neal** [comment]
Let me add these test vectors to double check on this impl

File: packages/credentials/src/utils.ts [hunk 1]
```diff
@@ -16,6 +16,33 @@ export function getCurrentXmlSchema112Timestamp(): string {
...
+ * ```
+ *
+ * @returns A date-time string in the format "yyyy-MM-ddTHH:mm:ssZ", compliant with XML Schema 1.1.2, based on the provided UNIX timestamp.
+ */
+export function getXmlSchema112Timestamp(nbf: number): string {
+  // Convert nbf from seconds to milliseconds and create a new Date object
+  const date = new Date(nbf * 1000);
+
+  // Format the date to an ISO string and then remove milliseconds
+  const issuanceDate = date.toISOString().replace(/\.\d{3}/, '');
```
Comment (decentralgabe) [inline packages/credentials/src/utils.ts:41]: this is correct, just calling out that in other libs we use RFC3339 - we may want to validate that both work against our test vectors cc: @mistermoe @mihai-chiorean

File: packages/credentials/src/verifiable-credential.ts [hunk 2]
```diff
@@ -194,17 +212,72 @@ export class VerifiableCredential {
...
+      throw new Error('Verification failed: exp claim does not match expirationDate');
+    }
+
+    // If exp is present, the UNIX timestamp MUST be converted to an [XMLSCHEMA11-2] date-time, and MUST be used to set the value of the expirationDate property of credentialSubject of the new JSON object.
+    if(exp) {
+      vcTyped.expirationDate = getXmlSchema112Timestamp(exp);
+    }
+
+    // TODO: iss is optional?
+    if (!iss) throw new Error('Verification failed: iss claim is required');
```
Comment (decentralgabe) [inline packages/credentials/src/verifiable-credential.ts:234]: not optional!

File: packages/credentials/src/verifiable-credential.ts [hunk 3]
```diff
@@ -194,17 +212,72 @@ export class VerifiableCredential {
...
+      throw new Error('Verification failed: sub claim does not match credentialSubject.id');
+    }
+
+    // If sub is present, the value MUST be used to set the value of the id property of credentialSubject of the new credential JSON object.
+    if(sub && !Array.isArray(vcTyped.credentialSubject)) {
+      vcTyped.credentialSubject.id = sub;
+    }
+
+    // jti MUST represent the id property of the verifiable credential or verifiable presentation.
+    if(jti && jti !== vcTyped.id) {
```
Comment (decentralgabe) [inline packages/credentials/src/verifiable-credential.ts:267]: what if there's no jti? shoudl we err?

**Review by decentralgabe — APPROVED** [review]
complex stuff, re-approving. nice job @nitro-neal

File: packages/credentials/src/verifiable-credential.ts (hunk 2 above)
Comment (nitro-neal) [reply packages/credentials/src/verifiable-credential.ts:234]: lol ok, just saw this: - https://www.w3.org/TR/vc-data-model/#jwt-decoding
```
If iss is present, the value MUST be used to set the issuer [property](https://www.w3.org/TR/vc-data-model/#dfn-property) of the new [credential](https://www.w3.org/TR/vc-data-model/#dfn-credential) JSON object or the holder [property](https://www.w3.org/TR/vc-data-model/#dfn-property) of the new [presentation](https://www.w3.org/TR/vc-data-model/#dfn-presentations) JSON object.
```

File: packages/credentials/src/utils.ts (hunk 1 above)
Comment (nitro-neal) [reply packages/credentials/src/utils.ts:41]: ok but if we accept a vc with RFC3339 we are not VcDataModel spec compliant yea?

File: packages/credentials/src/verifiable-credential.ts (hunk 3 above)
Comment (nitro-neal) [reply packages/credentials/src/verifiable-credential.ts:267]: image

so it seems like all of them are optional except for iss

File: packages/credentials/src/utils.ts (hunk 1 above)
Comment (nitro-neal) [reply packages/credentials/src/utils.ts]: I updated it accept RFC3339

**Comment by nitro-neal** [comment]
@mistermoe I added your test vectors - https://github.com/TBD54566975/web5-spec/tree/main/test-vectors/vc_jwt

**Review by frankhinek — APPROVED** [review]

**Pull request merged**